pkg/proctree/proctree.go

```diff
@@ -0,0 +1,118 @@
+package proctree
+
+import (
+	"os"
+
+	"github.com/aquasecurity/tracee/pkg/errfmt"
+	"github.com/aquasecurity/tracee/pkg/logger"
+	"github.com/aquasecurity/tracee/pkg/proctree/nodes"
+	"github.com/aquasecurity/tracee/pkg/proctree/tree"
+	"github.com/aquasecurity/tracee/pkg/utils"
+)
+
+type ProcTree struct {
+	tree *tree.Tree
+}
+
+func NewProcTree() *ProcTree {
+	// TODO: read from ProcFS from time to time
+
+	procTree := &ProcTree{
+		tree: tree.NewTree(),
+	}
+
+	err := procTree.UpdateFromProcFS()
+	if err != nil {
+		logger.Errorw("proctree procfs update", "error", err)
+	}
+
+	return procTree
+}
+
+//
+// All functions below can populate the tree somehow
+//
+
+// ParseFork handles a SchedProcessFork event
+func (p *ProcTree) ParseFork(given nodes.ProcessForkData) {
+	// Calculate hashes
+	childHash := utils.HashU32AndU64(uint32(given.Child.Pid), given.Child.StartTime)
+	parentHash := utils.HashU32AndU64(uint32(given.Parent.Pid), given.Parent.StartTime)
+	// Child
+	childProcess := p.tree.GetOrCreateProcess(childHash)
+	childProcess.SetInfo(given.Child)
+	// Parent
+	parentProcess := p.tree.GetOrCreateProcess(parentHash)
+	parentProcess.SetInfo(given.Parent)
+	// Link using unique identifiers (no need for transactional lock, parent change goes first)
+	parentProcess.AddChildHash(childHash)
+	childProcess.SetParentHash(parentHash)
+	// Debug (Uncomment this to see a tree representation using pretty ascii table) each fork
+	// fmt.Printf("%s", p.tree)
```

Comment:
Uncommenting this you will get a tree representation each time a fork happens:

![image](https://github.com/aquasecurity/tracee/assets/7395852/c94f783c-fb73-496c-9144-2826a7186b6e)

Of course just for debugging purposes. The missing information is likely related to existing processes (thus the WIP about reading /proc directory to feed the process tree).

Reply:
And yes there won't be zeros =), just testing.

Reply:
Awesome ❤️